// binance-api-client/src/models/websocket/ticker/TickerUpdate.ts
/**
 * Represents a single 24-hr ticker statistics collection.
 */
export class TickerUpdate {

    private _eventType: string;
    private _eventTime: Date;
    private _symbol: string;
    private _priceChange: number;
    private _priceChangePercentage: number;
    private _weightedAveragePrice: number;
    private _priorWindowTradePrice: number;
    private _lastPrice: number;
    private _lastQuantity: number;
    private _bidPrice: number;
    private _bidQuantity: number;
    private _askPrice: number;
    private _askQuantity: number;
    private _openingPrice: number;
    private _highestPrice: number;
    private _lowestPrice: number;
    private _baseVolume: number;
    private _quoteVolume: number;
    private _openingTime: Date;
    private _closureTime: Date;
    private _firstTradeId: number;
    private _lastTradeId: number;
    private _tradesCount: number;

    constructor( json: any ) {

        this._eventType = json.e;
        this._eventTime = new Date(json.E);
        this._symbol = json.s;
        this._priceChange = json.p;
        this._priceChangePercentage = json.P;
        this._weightedAveragePrice = json.w;
        this._priorWindowTradePrice = json.x;
        this._lastPrice = json.c;
        this._lastQuantity = json.C;
        this._bidPrice = json.b;
        this._bidQuantity = json.B;
        this._askPrice = json.a;
        this._askQuantity = json.A;
        this._openingPrice = json.o;
        this._highestPrice = json.h;
        this._lowestPrice = json.l;
        this._baseVolume = json.v;
        this._quoteVolume = json.q;
        this._openingTime = new Date( json.O );
        this._closureTime = new Date( json.C );
        this._firstTradeId = json.F;
        this._lastTradeId = json.L;
        this._tradesCount = json.n;

    }

    get eventType(): string {
        return this._eventType;
    }

    get eventTime(): Date {
        return this._eventTime;
    }

    get symbol(): string {
        return this._symbol;
    }

    get priceChange(): number {
        return this._priceChange;
    }

    get priceChangePercentage(): number {
        return this._priceChangePercentage;
    }

    get weightedAveragePrice(): number {
        return this._weightedAveragePrice;
    }

    get priorWindowTradePrice(): number {
        return this._priorWindowTradePrice;
    }

    get lastPrice(): number {
        return this._lastPrice;
    }

    get lastQuantity(): number {
        return this._lastQuantity;
    }

    get bidPrice(): number {
        return this._bidPrice;
    }

    get bidQuantity(): number {
        return this._bidQuantity;
    }

    get askPrice(): number {
        return this._askPrice;
    }

    get askQuantity(): number {
        return this._askQuantity;
    }

    get openingPrice(): number {
        return this._openingPrice;
    }

    get highestPrice(): number {
        return this._highestPrice;
    }

    get lowestPrice(): number {
        return this._lowestPrice;
    }

    get baseVolume(): number {
        return this._baseVolume;
    }

    get quoteVolume(): number {
        return this._quoteVolume;
    }

    get openingTime(): Date {
        return this._openingTime;
    }

    get closureTime(): Date {
        return this._closureTime;
    }

    get firstTradeId(): number {
        return this._firstTradeId;
    }

    get lastTradeId(): number {
        return this._lastTradeId;
    }

    get tradesCount(): number {
        return this._tradesCount;
    }

    set eventType(value: string) {
        this._eventType = value;
    }

    set eventTime(value: Date) {
        this._eventTime = value;
    }

    set symbol(value: string) {
        this._symbol = value;
    }

    set priceChange(value: number) {
        this._priceChange = value;
    }

    set priceChangePercentage(value: number) {
        this._priceChangePercentage = value;
    }

    set weightedAveragePrice(value: number) {
        this._weightedAveragePrice = value;
    }

    set priorWindowTradePrice(value: number) {
        this._priorWindowTradePrice = value;
    }

    set lastPrice(value: number) {
        this._lastPrice = value;
    }

    set lastQuantity(value: number) {
        this._lastQuantity = value;
    }

    set bidPrice(value: number) {
        this._bidPrice = value;
    }

    set bidQuantity(value: number) {
        this._bidQuantity = value;
    }

    set askPrice(value: number) {
        this._askPrice = value;
    }

    set askQuantity(value: number) {
        this._askQuantity = value;
    }

    set openingPrice(value: number) {
        this._openingPrice = value;
    }

    set highestPrice(value: number) {
        this._highestPrice = value;
    }

    set lowestPrice(value: number) {
        this._lowestPrice = value;
    }

    set baseVolume(value: number) {
        this._baseVolume = value;
    }

    set quoteVolume(value: number) {
        this._quoteVolume = value;
    }

    set openingTime(value: Date) {
        this._openingTime = value;
    }

    set closureTime(value: Date) {
        this._closureTime = value;
    }

    set firstTradeId(value: number) {
        this._firstTradeId = value;
    }

    set lastTradeId(value: number) {
        this._lastTradeId = value;
    }

    set tradesCount(value: number) {
        this._tradesCount = value;
    }
}
